Finish Week 4 Assignment and Start Week 5

sarahz916 opened the pull request:
Finished Week 4 Assignment:
  - Added maps API to travels.html with Spain map
  - Sentiment Analysis score displayed when comments are posted
  - email of logged in user is displayed when comments are posted
  - added some "nice-to-haves" on many portfolio files, might do them later

Started Week 5 Assignment:
  -intro to week 5 is modifying greeter file to pass greeter tests
  -wrote working FindMeetingQuery.java that passes FindMeetingQueryTests.java

Commit:
add sentiment analysis to comments some bugs still

Commit:
fix bug and sentiment analysis works for comments

Commit:
add map of Spain to travels, start week 5 assignment

Commit:
pass FindMeetingQueryTests

query commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -122,4 +128,16 @@ private String getUserEmail(){
+        LanguageServiceClient languageService = LanguageServiceClient.create();
+        Sentiment sentiment = languageService.analyzeSentiment(doc).getDocumentSentiment();
+        float score = sentiment.getScore();
+        System.out.println(score);
```
I assume this `println` was accidentally left here after debugging?

query commented on portfolio/src/main/webapp/index.html:
```diff
@@ -8,6 +8,9 @@
   </head>
+  <!-- Would be nice to be able to synchronize navbar accross all html files and not have to state is every time. -->
+  <!-- BUG: navbar does not display correctly on smaller width screens.-->
+  <!-- Nice to have: the navbar not disappear when scrolling down.-->
```
Take a look at [`position: sticky`](https://developer.mozilla.org/en-US/docs/Web/CSS/position).

query commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -14,10 +14,94 @@
+
+  /**Returns true if at least one attendee in Meeting Request is an attendee in the Event */
+  private boolean overlappingAttendees(MeetingRequest request, Event event){
+      for (String attendee : event.getAttendees()){
+          if (request.getAttendees().contains(attendee))
+            return true;
+        }
+      return false;
```
You can reduce this to a single operation using [`Collection#retainAll`](https://docs.oracle.com/javase/8/docs/api/java/util/Collection.html#retainAll-java.util.Collection-):

```java
return !request.getAttendees().retainAll(event.getAttendees()).isEmpty();
```

query commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -14,10 +14,94 @@
+      return false;
+  }
+
+  private boolean durationCheck (long duration, TimeRange timerange){
```
This function name isn't very clear. What about the duration is it checking?

query reviewed (changes requested):
It'd be nice to have some tests for `FindMeetingQuery`, in keeping with the TDD theme.

Commit:
add more test for optional attendees

sarahz916 committed:
Update DataServlet.java

Commit:
fix navbar to have sticky position

Commit:
fix navbar to have sticky position

sarahz916 committed:
Update FindMeetingQuery.java

sarahz916 committed:
Update FindMeetingQuery.java

sarahz916 committed:
Update FindMeetingQuery.java

sarahz916 commented:
Fixed Roger's suggestions (they were all super helpful!). 

Initial tests for FindingMeetingQuery were provided in a test folder already, but I also added some test so the file will show up in changed files.

Commit:
fix boolean cannot be dereferenced error

Commit:
add map to travel

Commit:
Add Runtime analysis

Commit:
fix navbar bug, now explands on narrow devices

Commit:
embed resume into resume page

Commit:
link logo to homepage

kkincy commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -122,4 +128,15 @@ private String getUserEmail(){
       }  
   }
+
+  /** Returns sentiment ananlysis value of comment a float between -1 and 1 */
```
Nit: "analysis" is misspelled. Also, can you clarify the meaning of the float returned in terms of sentiment?

Another nit: if it's a float, it should be between -1.0 and 1.0, and you can specify if it's inclusive or exclusive of those values.

kkincy commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -122,4 +128,15 @@ private String getUserEmail(){
+  private float getSentimentScore(String message) throws IOException {
+        Document doc =
+            Document.newBuilder().setContent(message).setType(Document.Type.PLAIN_TEXT).build();
+        LanguageServiceClient languageService = LanguageServiceClient.create();
```
Since this is a language-related service, does it default to English? Can you pass in other languages? It's best to specify the language in the comments.

kkincy commented on portfolio/src/main/webapp/index.html:
```diff
@@ -7,9 +7,13 @@
-  <nav class="navbar navbar-expand-lg navbar-light bg-light">
-  <a class="navbar-brand" href="#">Sarah Zou</a>
+  <!-- Would be nice to be able to synchronize navbar accross all html files and not have to state is every time. -->
+  <nav class="navbar sticky-top navbar-expand-lg navbar-light bg-light">
```
What do you mean by "state is"?

sarahz916 replied:
I meant state it. I think it's a framework thing. I'll look into it more this week, would make capstone project design easier.

kkincy commented on walkthroughs/week-5-tdd/intro/src/test/java/com/google/sps/GreeterTest.java:
```diff
@@ -30,4 +30,14 @@ public void testGreeting() {
+public void testGreetingTrimsWhitespace() {
+  Greeter greeter = new Greeter();
+
+  String greeting = greeter.greet("   Ada   ");
```
If the |greet| method only trims whitespace, you might want to rename it accordingly. Otherwise, if you add more functionality, it won't be as easy to unit test each of the features within |greet|. That is, more modularity might be useful here if you're planning on adding a lot to |greet|.

kkincy commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -14,10 +14,122 @@
+  // Ideally this problem could be dealt in O(n) time with the events sorted before 
+  // finding timeslots. 
+  public Collection<TimeRange> query (Collection<Event> events, MeetingRequest request){
+        // For optional attendees, just run query twice. First with optional  attendees included, if that is 
```
Nit: the indentation of these comments looks off.

kkincy commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -14,10 +14,122 @@
+    ArrayList<TimeRange> solutions = new ArrayList<>();
+    // No options for too long of a requests.
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()){
+      return solutions;
```
Rather than returning |solutions|, is it possible to return an error or log a warning? Otherwise, it might be worth commenting that |solutions| is returned empty upon failure.

kkincy commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -14,10 +14,122 @@
+    for (Event event: events){
+        // Check if event attendees are in the request. If not continue to next event.
+        boolean relevantEvent = overlappingAttendees(request, event);
+        if (!relevantEvent){
```
if (!overlappingAttendees(request, event))

(if my Java syntax isn't too rusty)

kkincy commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -14,10 +14,122 @@
+  }
+
+  //Runtime: O(n^2) where n is the size of the events. 
+  private Collection<TimeRange> queryHelper(Collection<Event> events, MeetingRequest request) {
```
What exactly is |queryHelper| doing here? Just removing invalid time ranges? Maybe it should be renamed to |removeInvalidTimeRanges| or |removeInvalidTimes|?

sarahz916 replied:
queryhelper was actually the original query function before we were asked to consider optional attendees.

kkincy commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -14,10 +14,122 @@
+
+  /**Splits a TimeRange into two time or shotens one TimeRange due to conflict with a given event */
+  // Runtime: O(n) where n is the size of the Old Solutions.
+  private ArrayList<TimeRange> takeOutConflicts(ArrayList<TimeRange> oldSolutions, Event event, long duration) {
```
If |takeOutConflicts| is only being called once, I'm wondering why it's a separate function rather than being the bulk of the logic within |queryHelper|, which has the more ambiguous function name.

kkincy commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -14,10 +14,122 @@
+    //make sure to check end point of time range is inclusive. 
+  }
+
+  /**Splits a TimeRange into two time or shotens one TimeRange due to conflict with a given event */
```
Nit: "shotens" -> "shortens"

Also, why is this splitting or shortening TimeRanges? What's the purpose of this method?

kkincy commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -14,10 +14,122 @@
+      ArrayList<TimeRange> solutions = new ArrayList<>();
+      for (TimeRange slot: oldSolutions){
+          if (slot.contains(event.getWhen())){
+              //need to split slot into two
```
Again, why do we need to split?

kkincy commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -14,10 +14,122 @@
+              //need to split slot into two
+              TimeRange slotA = TimeRange.fromStartEnd(slot.start(), event.getWhen().start(), false);
+              TimeRange slotB = TimeRange.fromStartEnd(event.getWhen().end(), slot.end(), false);
+              //check if duration is still good for both
```
For both what?

query commented:
> Initial tests for FindingMeetingQuery were provided in a test folder already, but I also added some test so the file will show up in changed files.

Ah, sorry, my mistake. Good to know.

query commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -93,15 +118,18 @@ else if (event.getWhen().end() > slot.start()){
-        }
-      return false;
+      // Since request.getAttendees returns an unmuttable set. Need to copy it.
+      Collection<String> requestedAttendees = new ArrayList<>();
+      requestedAttendees.addAll(request.getAttendees());
```
You can also just use the copy constructor:

```java
ArrayList<String> requestedAttendees = new ArrayList<>(request.getAttendees());
```

(I didn't catch that `#retainAll` mutated the collection in place. My mistake.)

query commented on walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:
```diff
@@ -270,5 +272,125 @@ public void notEnoughRoom() {
+        new Event("Event 1", TimeRange.fromStartDuration(TIME_0800AM, DURATION_30_MINUTES),
+            Arrays.asList(PERSON_A)),
+        new Event("Event 2", TimeRange.fromStartDuration(TIME_0900AM, DURATION_30_MINUTES),
+            Arrays.asList(PERSON_B)));
+
+    MeetingRequest request =
+        new MeetingRequest(Arrays.asList(), DURATION_30_MINUTES);
+    request.addOptionalAttendee(PERSON_A);
+    request.addOptionalAttendee(PERSON_B);
+
+    Collection<TimeRange> actual = query.query(events, request);
+    Collection<TimeRange> expected =
+        Arrays.asList(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TIME_0800AM, false),
+            TimeRange.fromStartEnd(TIME_0830AM, TIME_0900AM, false),
+            TimeRange.fromStartEnd(TIME_0930AM, TimeRange.END_OF_DAY, true));
+
+    Assert.assertEquals(expected, actual);
+  }
+
+  @Test
+  public void noOptionsForOptionalAttendeesOnly() {
+
+    Collection<Event> events = Arrays.asList(
+        new Event("Event 1", TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TIME_1100AM, false),
+            Arrays.asList(PERSON_A)),
+        new Event("Event 2", TimeRange.fromStartEnd(TIME_1100AM, TimeRange.END_OF_DAY, false),
+            Arrays.asList(PERSON_B)));
+
+    MeetingRequest request =
+        new MeetingRequest(Arrays.asList(), DURATION_30_MINUTES);
+    request.addOptionalAttendee(PERSON_A);
+    request.addOptionalAttendee(PERSON_B);
+
+    Collection<TimeRange> actual = query.query(events, request);
+    Collection<TimeRange> expected =
+        Arrays.asList();
+
+    Assert.assertEquals(expected, actual);
+  }
+
```
👍

kkincy commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -14,10 +14,122 @@
+  /**Returns true if at least one attendee in Meeting Request is an attendee in the Event */
+  // Runtime: addAll and retainAll are O(n) time (I could be wrong on that). 
+  private boolean overlappingAttendees(MeetingRequest request, Event event){
+      // Since request.getAttendees returns an unmuttable set. Need to copy it.
```
"unmuttable" - > "immutable"

kkincy commented on walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:
```diff
@@ -270,5 +272,125 @@ public void notEnoughRoom() {
+    request.addOptionalAttendee(PERSON_C);
+    Collection<TimeRange> actual = query.query(events, request);
+    Collection<TimeRange> expected =
+        Arrays.asList(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TIME_0800AM, false),
```
What is false here? Can you add an in-line comment here and elsewhere?

sarahz916 replied:
false means endtime is not included in the time range.

sarahz916 committed:
Update FindMeetingQuery.java

Commit:
addressing Karen's comments
